enhancements/baremetal/baremetal-keepalived-unicast.md

```diff
@@ -0,0 +1,179 @@
+
+As mentioned above, `api-vip` and `ingress-vip` are supported to enable high availability.
+
+The `api-vip` should resides on the bootstrap instance during the bootstrap phase and move to one of the control plane nodes after the bootstrap process has completed and the bootstrap instance is stopped.
+The group members for the `api-vip` should include the bootstrap node and the control plane nodes.
+
+As per the `ingress-vip`, it should resides on a node (either control or compute) that runs a router pod, so for that case the group members should include all control and compute nodes.
+
+#### api-vip implementation
+
+Since the bootstrap IP address is not known until the server is up, a method for 
+retrieving that node will need to be implemented to allow the control plane nodes to include the 
+bootstrap IP address in their Keepalived configuration. 
+
+##### Bootstrap node
+
+The bootstrap [Keepalived](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml) instance runs as static pod and the relevant [Keepalived configuration](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl) rendered by the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg).
+
+The bootstrap Keepalived pod includes:
+- [Keepalived container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L35-#L93)
+- [TCP server container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L27#L34) that will be responsible to return the bootstrap IP address.
+- [keepalived-monitor container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L94-#L138) is responsible to render the Keepalived config file.
+
+The keepalived-monitor container in bootstrap node will first render Keepalived cfg file with an empty 'unicast_peer' section, as a result of that the `api-vip` will be set to the bootstrap node by Keepalived.
+
+At the next step, the bootstrap's keepalived-monitor container will repeatedly retrieve the control plane nodes details from the localhost:kube-apiserver and add them to 'unicast_peer' section.
+
+The bootstrap node owns the `api-vip' during bootstrap phase because it's configured with [higher](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl#L13) VRRP priority than the control plane nodes [priority](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L26).
```

Comment:
```suggestion
The bootstrap node owns the `api-vip` during bootstrap phase because it's configured with [higher](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl#L13) VRRP priority than the control plane nodes [priority](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/templates/master/00-master/baremetal/files/baremetal-keepalived-keepalived.yaml#L26).
```

Reply:
Done